### PR TITLE
Circuit drone + grabber/thrower balance changes

### PR DESCRIPTION
**Pull request by evsey9**
[Changelogs]: 
:cl: Evsey9
balance: Integrated Circuit Drones are now bulky to improve safety ratings, and therefore, cannot be stored in normal backpacks or pockets, but now can accept modules that bulky circuit machinery can.
balance: Grabbers can now grab items up to the size of the circuit assembly they are in.
balance: Grabbers cannot store circuit machinery the same or larger size than the circuit assembly they are in.
tweak: Throwers now can throw items up to the size of circuit assembly they are in.
/:cl:
tl;dr drones are now bulky items, grabbers max_w_class is w_class of circuit assembly they're in, grabbers cannot hold same assemblies that they're in ( to prevent infinite stacking )
[why]: 1. It makes no sense that you can fit devices which can hold multiple guns, grenades, themselves etc. into POCKETS and backpacks. There is basically no reason to use any other assemblies since this is better than the circuit machine one, yet smaller and fits more into it..?
2. This should make so that you can't stack annoying drones which scream stuff constantly inside your backpack and run around.
3. Allowing bulky items to be stored seems like a _somewhat_ fair tradeoff for the two points I mentioned above.

This probably has bigger balance implications than what I thought of, so, criticism/review and such appreciated.

### CODE REVIEW

**Comment by kevinz000** [comment]
Lol guess we can't wait for me to do it instead have fun selea

**Comment by evsey9** [comment]
please elaborate ._.

**Comment by epochayur** [comment]
does this apply to all IC bots? makes sense for the ones that can grab/pull, but there's no reason the smallest bot shouldn't be able to fit in a backpack

**Comment by evsey9** [comment]
Integrated circuit DRONES. The ones with the arms, that can move with loco circuits

**Comment by epochayur** [comment]
alright sure then

**Comment by Repukan** [comment]
Not sure about the bulky item grabbers. Seems like it'd just facilitate bot voice tracker remote maxcap bombing.

**Comment by epochayur** [comment]
wait, can you infinitely store bulky drones within bulky drones, like some sort of fucked up text-spamming russian doll?

**Comment by evsey9** [comment]
you could already do that with small drones
fun fact: drones can store things larger than they are

**Comment by epochayur** [comment]
@arsserpentarium

**Comment by evsey9** [comment]
if there's a weight class larger than bulky then I COULD fix it

**Comment by epochayur** [comment]
or just disallow drones from holding drones, electromagnetic interference something something

**Comment by evsey9** [comment]
yeah sure i can try doing that

**Comment by evsey9** [comment]
done
writing my first piece of actual code was quite a challenge

**Comment by Anonmare** [comment]
@evsey9 There is, exosuit equipment uses it

**Comment by arsserpentarium** [comment]
I think, that it's good balance change.I appreciate it.
About grabbers, maybe we should make smal/medium/large grabbers. With ability to store items lesser, than this grabber.I was going to make it, but forgot.
But maybe,there should be big drone and small drone.Not sure.

**Comment by arsserpentarium** [comment]
Why?Because not so many applications, where you need to store bulky and large items.Mostly you need to store small ones.

**Comment by kevinz000** [comment]
@evsey9 i was going to overhaul drones at the very least and probably lots of shit about assemblies but whateveri 'm lazy doing good work friend.

**Comment by Kerbin-Fiber** [comment]
In the Changelog description, it tells you that everything is changed to large but everything else says bulky.
Fix it.

**Comment by evsey9** [comment]
large and bulky are different weight classes?

**Comment by evsey9** [comment]
ok, change of plan. i'll just make drones be bulky in this PR and make so that grabber's max w_class is the w_class of the assembly

**Comment by arsserpentarium** [comment]
Grabber size is assembly size.
And it can store things with lesser size.
It will do the trick.

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 1]
```diff
@@ -304,19 +304,26 @@
...
+				var/is_circuit = istype(AM,/obj/item/device/electronic_assembly/)
+				var/circuitcheck = 0
+				if (!is_circuit)
+					circuitcheck = 1
+				else
+					if (AM.w_class < max_w_class)
+						circuitcheck = 1
+					else
+						circuitcheck = 0
+				if((contents.len < max_items) && (!max_w_class || AM.w_class <= max_w_class) && (circuitcheck))
```
[the remark this answers is not in the document]
Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: circuitcheck only checks for the w_class only if the thing is an integrated circuit, and in other cases it just returns a 1

**Comment by evsey9** [comment]
Should I make it so that thrower's max w_class would be dependant on w_class of the assembly? It would seem logical for me, buuut it might have bigger balance implications.

**Comment by arsserpentarium** [comment]
Well, I can't see, why it could be problem.
Turret, which can fire turrets is intended.Just without infinite nesting.

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 2]
```diff
@@ -304,19 +304,28 @@
...
+	var/max_w_class = assembly.w_class
 	var/atom/movable/acting_object = get_object()
 	var/turf/T = get_turf(acting_object)
 	var/obj/item/AM = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
 	if(AM)
 		var/mode = get_pin_data(IC_INPUT, 2)
-
 		if(mode == 1)
 			if(check_target(AM))
-				if((contents.len < max_items) && (!max_w_class || AM.w_class <= max_w_class))
```
Comment (arsserpentarium) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: if((contents.len < max_items) && (!max_w_class || AM.w_class < max_w_class))
        AM.forceMove(src)

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 3]
```diff
@@ -408,9 +417,9 @@
 	)
 	spawn_flags = IC_SPAWN_RESEARCH
 	power_draw_per_use = 50
-	var/max_w_class = WEIGHT_CLASS_NORMAL
 
 /obj/item/integrated_circuit/manipulation/thrower/do_work()
+	var/max_w_class = assembly.w_class
```
Comment (arsserpentarium) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: change line if(max_w_class && (A.w_class > max_w_class))
to if(max_w_class && (A.w_class >= max_w_class))

Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: it's intended for the thrower to be able to throw items the size of the assembly, though? as in, bulky assemblies throw bulky things, normal assemblies throw normal stuff, small assemblies throw small stuff

Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: well okay.

Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: Honestly, at first it was a bug.But it was so hilarious, so I decided to leave it.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: Remember, the creeper was also a bug and turned into the face of the entire game it was made for!

**Comment by awesine** [comment]
have  a series of bulky drone grabbing bulky drones grabbing bulky drones .etc .etc .etc so like 100 drones are all on one tile and then activate their sound circuits

:^)

**Comment by evsey9** [comment]
I actally fixed the infinite circuit grabbing, thats what the weightcheck and istype there for. But hey, if another drone gets near another - they can throw eachother!

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 4]
```diff
@@ -304,19 +304,28 @@
 	activators = list("pulse in" = IC_PINTYPE_PULSE_IN,"pulse out" = IC_PINTYPE_PULSE_OUT)
 	spawn_flags = IC_SPAWN_RESEARCH
 	power_draw_per_use = 50
-	var/max_w_class = WEIGHT_CLASS_NORMAL
 	var/max_items = 10
-
 /obj/item/integrated_circuit/manipulation/grabber/do_work()
+	var/max_w_class = assembly.w_class
```
Comment (kevinz000) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: Allows ttv grabbing and fully bypassing the ttv nerf as almost no one is going to search every drone.

Comment (kevinz000) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: @KorPhaeron

Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: last time i tried blowing up TTVs inside of drones nothing happened ( they nullspaced or something )

Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: ok it should no longer be able to grab TTVs

File: code/modules/integrated_electronics/subtypes/manipulation.dm [hunk 5]
```diff
@@ -304,19 +304,28 @@
...
+					if (AM.w_class <= max_w_class)
+						weightcheck = TRUE
+					else
+						weightcheck = FALSE
+				else
+					if (AM.w_class < max_w_class)
+						weightcheck = TRUE
+					else
+						weightcheck = FALSE
+				if((contents.len < max_items) && (weightcheck))
```
Comment (arsserpentarium) [inline code/modules/integrated_electronics/subtypes/manipulation.dm]: just leave there 
if((contents.len < max_items) && (!max_w_class || AM.w_class < max_w_class))
Large objects still sufficient. But better to have little more restrictions, than hardcoded list of restricted items.

Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: what the FUCK are large items? can i get at least one example of a large object that ISN'T just a normal sized object

Comment (evsey9) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: also, that would mean grabbers in normal sized circuit would be able to only store small items, and grabbers in small sized circuit wouldn't be able tto store anything at all, which i'm not really for

Comment (arsserpentarium) [reply code/modules/integrated_electronics/subtypes/manipulation.dm]: oh fuck, you are right.
BTW, now you can't even stick grabber into small circuit assembly.
Which is fully logical.

**Comment by BlueNothing** [comment]
What's the difference between drone assemblies and the other assemblies, again?

**Comment by evsey9** [comment]
Drones are the only assemblies that can move

**Comment by BlueNothing** [comment]
So the other ones are stuck into the floor like a protolathe or a comms
console?

**Comment by evsey9** [comment]
Im really not in the mood for explaining to you how circuits work but they're items that you can carry, drones can use locomotion circuits which let them move around by itself

**Comment by BlueNothing** [comment]
Sorry, I derped.

**Pull request merged**